homeassistant/components/weatherkit/coordinator.py

```diff
@@ -0,0 +1,75 @@
+"""DataUpdateCoordinator for WeatherKit integration. Updated every 15 minutes."""
+from __future__ import annotations
+
+from datetime import timedelta
+from typing import Literal
+
+from apple_weatherkit.client import (
+    WeatherKitApiClient,
+    WeatherKitApiClientAuthenticationError,
+    WeatherKitApiClientError,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_LATITUDE, CONF_LOCATION, CONF_LONGITUDE
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, LOGGER
+
+
+class WeatherKitDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching data from the API."""
+
+    config_entry: ConfigEntry
+    supported_data_sets: list[
+        Literal["currentWeather", "forecastDaily", "forecastHourly"]
+    ] | None
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        client: WeatherKitApiClient,
+    ) -> None:
+        """Initialize."""
+        self.client = client
+        self.supported_data_sets = None
+        super().__init__(
+            hass=hass,
+            logger=LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(minutes=15),
+        )
+
+    async def _update_supported_data_sets(self):
+        supported_data_sets = await self.client.get_availability(
+            self.config_entry.data[CONF_LOCATION][CONF_LATITUDE],
+            self.config_entry.data[CONF_LOCATION][CONF_LONGITUDE],
+        )
+
+        requested_data_sets = ["currentWeather", "forecastDaily", "forecastHourly"]
+
+        self.supported_data_sets = [
+            data_set
+            for data_set in requested_data_sets
+            if data_set in supported_data_sets
+        ]
+
+        LOGGER.debug("Supported data sets: %s", self.supported_data_sets)
+
+    async def _async_update_data(self):
+        """Update the current weather and forecasts."""
+        try:
+            if not self.supported_data_sets:
+                await self._update_supported_data_sets()
+
+            return await self.client.get_weather_data(
+                self.config_entry.data[CONF_LOCATION][CONF_LATITUDE],
+                self.config_entry.data[CONF_LOCATION][CONF_LONGITUDE],
+                self.supported_data_sets,
+            )
+        except WeatherKitApiClientAuthenticationError as exception:
+            raise ConfigEntryAuthFailed(exception) from exception
```

Comment:
This would trigger a reauth flow, but you don't have any. Can we check if we can actually get data in the `__init__.py`'s `async_setup_entry` and raise the ConfigEntryAuthFailed there?

Reply:
What is the difference throwing it there? Not triggering the re-auth flow?

Reply:
Good question, not sure atm

Reply:
Either method is going to trigger the reauth flow which will fail since it doesn't exist.

You can manually test the credentials in `__init__.py:async_setup_entry`, log an error, and return `False` if it fails for now.

In a future PR, change it to `raise ConfigEntryAuthFailed`

Also there is no need to call `ConfigEntryAuthFailed(exception)` when you do `raise ConfigEntryAuthFailed from exception` since it will figure out how to handle it from the cause